[lldb] Turn lldb_private::Status into a value type.

adrian-prantl opened the pull request:
This patch removes all of the Set.* methods from Status.

This cleanup is part of a series of patches that make it harder use the anti-pattern of keeping a long-lives Status object around and updating it while dropping any errors it contains on the floor.

This patch is largely NFC, the more interesting next steps this enables is to:
1. remove Status.Clear()
2. assert that Status::operator=() never overwrites an error
3. remove Status::operator=()

Note that step (2) will bring 90% of the benefits for users, and step (3) will dramatically clean up the error handling code in various places. In the end my goal is to convert all APIs that are of the form

`    ResultTy DoFoo(Status& error)
`
to

`    llvm::Expected<ResultTy> DoFoo()
`
How to read this patch?

The interesting changes are in Status.h and Status.cpp, all other changes are mostly

`     perl -pi -e 's/\.SetErrorString/ = Status::FromErrorString/g' $(git grep -l SetErrorString lldb/source)
`
plus the occasional manual cleanup.

medismailben commented on lldb/include/lldb/Utility/Status.h:
```diff
@@ -56,19 +56,41 @@ class Status {
+  static Status FromErrorString(const char *format) {
+    if (format)
+      return Status(std::string(format));
+    return Status(std::string("null error"));
```
```suggestion
    return Status(llvm::StringRef("null error"));
```

This should be more efficient.

adrian-prantl replied:
Note that Status stores a `std::string` internally and there's a `Status(std::string&&)` constructor at work here.

medismailben commented on lldb/test/API/functionalities/scripted_process/invalid_scripted_process.py:
```diff
@@ -15,7 +15,7 @@ def __init__(self, exe_ctx: lldb.SBExecutionContext, args: lldb.SBStructuredData
         self, addr: int, size: int, error: lldb.SBError
     ) -> lldb.SBData:
-        error.SetErrorString("This is an invalid scripted process!")
+        error = Status::FromErrorString("This is an invalid scripted process!")
```
This is not valid python. The regex should only be applied to C++.

adrian-prantl replied:
Thanks!

medismailben reviewed (commented):
Cool stuff! I think this is fine except for the python files.

clayborg reviewed (commented):
I worry the assert will kill the LLDB library and cause issues. Can we add a temporary setting that can override this in case it does cause crashes? I really don't want LLDB crashing if we can help it. It will be hard to test all of the error code paths that can happen out in the wild

Michael137 commented on lldb/source/Utility/Status.cpp:
```diff
@@ -51,13 +51,13 @@ Status::Status(std::error_code EC)
-  SetErrorStringWithVarArg(format, args);
-  va_end(args);
-}
+Status::Status(std::string &&err_str)
```
Why `std::string&&`? Looks like in a lot of the cases you're going to have to do a c-str -> std::string conversion anyway? For the `FromErrorStringWithFormat` where you construct `Status` with an l-value string, you're actually silently copying (because this `std::string&&` overload isn't picked, instead it picks the `llvm::StringRef` constructor, and copies the string).

All this might just be easier to reason about if you make it a by-value argument, for all cases. And the double-move (for std::move construction+std::move from the argument is probably not worth optimizing away by-hand?).

labath commented:
I like this. I have just two remarks:
- it might be better to split this into three steps (add new APIs, port to new APIs, remove old APIs), as that will make reverts easier/less disruptive (I don't know how much we can trust pre-commit CI these days, but I wouldn't be surprised if this breaks some platform-specific code).
- since this seems like a perfect opportunity to bikesh^Wdiscuss the names, I'm going to ask if there's any appetite for shortening some of the new factory functions. `Status::FromErrorStringWithFormatv` is a bit of a mouthful, so I was thinking if we could  use something shorter instead (`Status::FromFormatv` or even `Status::Formatv`) ?

adrian-prantl committed:
[lldb] Turn lldb_private::Status into a value type.

This patch removes all of the Set.* methods from Status.

This cleanup is part of a series of patches that make it harder use
the anti-pattern of keeping a long-lives Status object around and
updating it while dropping any errors it contains on the floor.

This patch is largely NFC, the more interesting next steps this enables is to:
1. remove Status.Clear()
2. assert that Status::operator=() never overwrites an error
3. remove Status::operator=()

Note that step (2) will bring 90% of the benefits for users, and step
(3) will dramatically clean up the error handling code in various
places. In the end my goal is to convert all APIs that are of the form

    ResultTy DoFoo(Status& error)

to

    llvm::Expected<ResultTy> DoFoo()

How to read this patch?

The interesting changes are in Status.h and Status.cpp,
all other changes are mostly

     perl -pi -e 's/\.SetErrorString/ = Status::FromErrorString/g' $(git grep -l SetErrorString lldb/source)

plus the occasional manual cleanup.

adrian-prantl commented:
> I worry the assert will kill the LLDB library and cause issues. Can we add a temporary setting that can override this in case it does cause crashes? I really don't want LLDB crashing if we can help it. It will be hard to test all of the error code paths that can happen out in the wild

@clayborg When I say assert I had in mind something like this:
```
Status &Status::operator=() {
  assert(IsSuccess() && "dropping an error Status without checking first");
  #if NDEBUG
  if (!IsSuccess())
    Debugger::ReportWarning("dropping an error Status without checking first");
  #endif
  ...
```

So it would only "crash" in an asserts build, but otherwise keep working in production. Does that address you concern?
I'm also not planning to change the contract of the SBAPI. You'll still be able to modify SBError without triggering an assertion. It will check and log m_status under the hood.

adrian-prantl commented:
> I like this. I have just two remarks:
> 
> * it might be better to split this into three steps (add new APIs, port to new APIs, remove old APIs), as that will make reverts easier/less disruptive (I don't know how much we can trust pre-commit CI these days, but I wouldn't be surprised if this breaks some platform-specific code).

From previous experience, I'm convinced that I'm going to break some platform-specific bots with this commit. I like the idea of only reverting the commit that removes the old API while that process is ongoing!

> * since this seems like a perfect opportunity to bikesh^Wdiscuss the names, I'm going to ask if there's any appetite for shortening some of the new factory functions. `Status::FromErrorStringWithFormatv` is a bit of a mouthful, so I was thinking if we could  use something shorter instead (`Status::FromFormatv` or even `Status::Formatv`) ?

I picked these names, because they are in line with the old names, which made the regex replacement feasible. Renaming them afterwards is going to be easier.
My 2 cents on the naming: I had `Status::FromFormatv` in a previous iteration of this patch and changed my mind, because it doesn't indicate that this is going to be an error. What do you think about the slightly shorter `Status::ErrorFromFromatv()`?
Or, more radical, and potentially *really confusing* with LLVM code: rename `lldb_private::Status` to `lldb_private::Error`. I don't think I'd like that.

jimingham commented:
> On Aug 27, 2024, at 8:51 AM, Adrian Prantl ***@***.***> wrote:
> 
> 
> I like this. I have just two remarks:
> 
> it might be better to split this into three steps (add new APIs, port to new APIs, remove old APIs), as that will make reverts easier/less disruptive (I don't know how much we can trust pre-commit CI these days, but I wouldn't be surprised if this breaks some platform-specific code).
> From previous experience, I'm convinced that I'm going to break some platform-specific bots with this commit. I like the idea of only reverting the commit that removes the old API while that process is ongoing!
> 
> since this seems like a perfect opportunity to bikesh^Wdiscuss the names, I'm going to ask if there's any appetite for shortening some of the new factory functions. Status::FromErrorStringWithFormatv is a bit of a mouthful, so I was thinking if we could use something shorter instead (Status::FromFormatv or even Status::Formatv) ?
> I picked these names, because they are in line with the old names, which made the regex replacement feasible. Renaming them afterwards is going to be easier.
> My 2 cents on the naming: I had Status::FromFormatv in a previous iteration of this patch and changed my mind, because it doesn't indicate that this is going to be an error. What do you think about the slightly shorter Status::ErrorFromFromatv()?
> Or, more radical, and potentially really confusing with LLVM code: rename lldb_private::Status to lldb_private::Error. I don't think I'd like that.
> 

Note that lldb::Status WAS lldb::Error for a very long time till LLVM added its Error classes, and renamed the lldb ones in doing so.   That's why you see

Status error;

everywhere in lldb.  So going back is unlikely to find favor...

Jim

> —
> Reply to this email directly, view it on GitHub <https://github.com/llvm/llvm-project/pull/106163#issuecomment-2312935884>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADUPVWYUU3EX3LF7GGFUCJ3ZTSOAPAVCNFSM6AAAAABNFAYGXKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGMJSHEZTKOBYGQ>.
> You are receiving this because you are on a team that was mentioned.
>

adrian-prantl commented:
Updated all the non-Darwin plugins. I believe to have addressed all outstanding comments now.

bulbazord reviewed (approved):
I haven't reviewed every single file but the ones I did look at look straightforward and correct. The changes to `Status` itself I feel positively about. I agree with the direction that this takes `Status` in.

clayborg commented:
My concerns are addressed, thanks @adrian-prantl

Michael137 commented on lldb/source/Utility/Status.cpp:
```diff
@@ -37,10 +37,10 @@ class raw_ostream;
-Status::Status(ValueType err, ErrorType type)
-    : m_code(err), m_type(type), m_string() {}
+Status::Status(ValueType err, ErrorType type, std::string msg)
+    : m_code(err), m_type(type), m_string(msg) {}
```
std::move (and below too)

JDevlieghere reviewed (approved):
This is great. It's something I've wanted since the first day I started working on LLDB. We've made a lot of progress in this direction so it's great to see that we've reached a point where we can make this actually happen.

labath commented:
> > * since this seems like a perfect opportunity to bikesh^Wdiscuss the names, I'm going to ask if there's any appetite for shortening some of the new factory functions. `Status::FromErrorStringWithFormatv` is a bit of a mouthful, so I was thinking if we could  use something shorter instead (`Status::FromFormatv` or even `Status::Formatv`) ?
> 
> I picked these names, because they are in line with the old names, which made the regex replacement feasible. Renaming them afterwards is going to be easier. My 2 cents on the naming: I had `Status::FromFormatv` in a previous iteration of this patch and changed my mind, because it doesn't indicate that this is going to be an error. What do you think about the slightly shorter `Status::ErrorFromFromatv()`?

I think it's better (because its shorter), though it's still a bit long for my taste. FWIW, I don't find the short name at all confusing, because my mental model of these error types is (and all types I know behave that way) is that they have only one "success" value, and everything else represents some kind of an error/failure. So, when I see anything more complicated than `Status()` (or whatever the final API is for constructing success values), I immediately expect an error status.

I guess one difference is that I'm used to working with `absl::Status` (which I believe was zturner's inspiration for this name), so "status" feels just like a synonym for "error" to me. If you think having the word "error" in the name helps, I'd also be fine with `Status::Errorf` and `Status::Errorv` (for printf vs formatv styles) or something like that.
